New: Legend Item Symbol Dashed Line

SharathAlur opened the pull request:
<!-- Thanks for submitting a pull request! Please provide enough information so that others can review your pull request. -->
<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->
<!-- Try to link to an open issue for more information. -->

**What is the purpose of this pull request? (Check any that's applicable)**
Add a line in the legend button along with the symbol.

-   [x] Feature
-   [ ] Bug fix
-   [ ] Dependency updates
-   [ ] Documentation updates
-   [ ] Build related changes
-   [ ] Changes an existing functionality
-   [ ] Other, please explain:

**Does this introduce a breaking change?**

-   [ ] Yes
-   [x] No

<!-- If this introduces a breaking change, please describe the impact and migration path for existing applications below. -->

**What changes did you make? (Give an overview)**
Added legendOptions to provide users, the ability to display shape and/or line in the legend button.

**Add any screenshots**
<img width="1460" alt="LegendOptions" src="https://user-images.githubusercontent.com/22268936/74756815-2f962900-529b-11ea-89bb-17ca87e87265.png">

pranav300 commented on docs/controls/Line.md:
```diff
@@ -193,6 +194,17 @@ Each line can have a style object in [Values](#values) level.
 
+##### LegendOptions
+
+Each line cna have a legendOptions object in [Values](#values) level.
```
Everything looks good to me. Just a minor change for `can`.

SharathAlur replied:
I have updated [here|https://github.com/SharathAlur/carbon-graphs/commit/b9324c1e209febc4aaffbdbd235f2a6812b10f5d]

erinplourde commented:
UX Feedback: 
**Overall:** What is the size of the over button? When toggled off does the bottom still present with the grey border and the white fill? 
**Symbol only:** The symbol looks big, are we able to being the size of that down? Also, it looks like the white stroke surrounding it it getting cut off around some of the edges. 
**Line/Symbol Combination:** In the example with the triangle and the line, the white borders are not touching between the symbol in and the line. It is preferable to have approx. 1 px white stroke between the line and the symbol. (See Below) Also, the white looks to be inconsistent around the line and the symbol. both should have a 1 px border of white.  Lastly, is the section with the line about 17 px long ?(that is including the the white border)
<img width="92" alt="Screen Shot 2020-02-25 at 8 36 54 AM" src="https://user-images.githubusercontent.com/52933964/75257019-fc2b3100-57a9-11ea-913d-697e096747f6.png">

**Line Only:** White border around the line looks to be inconsistent - should be 1 px stroke around the entire color line. The line is sitting too close the edges of the buttons - Is the line about 28 px?

SharathAlur commented:
@erinplourde 
**Overall:**
- The size of the button was 30*30. Now I have updated according to the draft. i.e., 30*35
<img width="278" alt="Button-size" src="https://user-images.githubusercontent.com/22268936/75321644-37575e00-5897-11ea-9e0e-4adc610b4213.png">
- When the button is toggled off, the button don't show any border. This is the existing functionality. I have not made any changes to this.
<img width="278" alt="Toggle-Button" src="https://user-images.githubusercontent.com/22268936/75321746-6ec60a80-5897-11ea-8a5f-423f9e90707a.png">
**Symbol only:**
- I missed updating the size of the icon. I have updated it to 10px now.
<img width="278" alt="Icon-only" src="https://user-images.githubusercontent.com/22268936/75321907-c95f6680-5897-11ea-9f39-a785d1672a71.png">
- Cutting edges were present previously. I have just modified the size as of now.
**Line Only:**
- I have corrected it to display the white border consistently around the line.
**Line/Symbol Combination:**
- Unresolved. Needs to be worked on.

SharathAlur commented:
@erinplourde 
**Line/Symbol Combination:** We are using separate svg for both symbol and line. Having svg in a square, there is a technical constraint to keep the line and the shape touching.
CC: @abhijit945

abhijit945 commented on dev/examples/data.js:
```diff
@@ -68,7 +72,11 @@ const DATA = [
+                    })),
+                    legendOptions: {
+                        showLine: true,
+                        showShape: true
```
I would add `style.strokeDashArray` here.

SharathAlur replied:
Added

abhijit945 commented on docs/controls/Line.md:
```diff
@@ -193,6 +194,17 @@ Each line can have a style object in [Values](#values) level.
+
+| Property Name | Expected | Default | Description                  |
+| ------------- | -------- | ------- | ---------------------------- |
+| showShape     | boolean  | true    | Display Shape in the legend. |
```
To maintain passivity we need to set this as `true` and showLine as `false`. Let consumers decide what they want to show/hide after that.

SharathAlur replied:
Updated

abhijit945 commented on src/main/js/controls/Line/helpers/helpers.js:
```diff
@@ -272,6 +272,10 @@ const processDataPoints = (graphConfig, dataTarget) => {
     dataTarget.style = {
         strokeDashArray: getStrokeDashArray(dataTarget.style)
     };
+    dataTarget.legendOptions = getDefaultValue(dataTarget.legendOptions, {
```
Can you move this to `helpers/legend.js`?
```js
const getDefaultLegendOptions = () => getDefaultValue(dataTarget.legendOptions, {
    showShape: getDefaultValue(graphConfig.showShapes, true),
    showLine: true,
   // ... others here
});
```

abhijit945 commented on src/main/js/helpers/legend.js:
```diff
@@ -92,27 +92,85 @@ const loadLegendItem = (legendSVG, t, shownTargets, eventHandlers) => {
+        }
+    } else {
+        displayIcon(button, t, false);
+    }
```
Move these decisions to another function. 
```js
const processLegendOptions = (options) => {
    if (options) {
        if (options.showShape) {
            displayIcon(buttonPath, options);
        }
        if (options.showLine) {
            displayLine(buttonPath, options);
        }
    } else {
        displayIcon(buttonPath);
    }
};
```

abhijit945 commented on src/main/js/helpers/legend.js:
```diff
@@ -92,27 +92,85 @@ const loadLegendItem = (legendSVG, t, shownTargets, eventHandlers) => {
+            displayLine(button, t, t.legendOptions.showShape);
+        }
+    } else {
+        displayIcon(button, t, false);
```
`buttonPath`

abhijit945 commented on src/main/js/helpers/legend.js:
```diff
@@ -92,27 +92,85 @@ const loadLegendItem = (legendSVG, t, shownTargets, eventHandlers) => {
+ * @param {object} dataPoint - Data points object
+ * @param {boolean} showLine - if the line is shown along with icon
+ */
+const displayIcon = (itemPath, dataPoint, showLine) => {
```
I would change this to `createLegendIcon`

abhijit945 commented on src/main/js/helpers/legend.js:
```diff
@@ -92,27 +92,85 @@ const loadLegendItem = (legendSVG, t, shownTargets, eventHandlers) => {
+ * @param {object} dataPoint - Data points object
+ * @param {object} showShape - Shape is shown along with line
+ */
+const displayLine = (itemPath, dataPoint, showShape) => {
```
`createLegendLine`

abhijit945 commented on src/main/js/helpers/legend.js:
```diff
@@ -92,27 +92,85 @@ const loadLegendItem = (legendSVG, t, shownTargets, eventHandlers) => {
+    svg.append("line") // creating white line
+        .attr("x1", 0)
+        .attr("x2", 35)
+        .attr("style", `stroke: ${COLORS.WHITE}; stroke-width: 6;`);
```
Can we use CSS for this?

abhijit945 commented on src/main/js/helpers/legend.js:
```diff
@@ -92,27 +92,85 @@ const loadLegendItem = (legendSVG, t, shownTargets, eventHandlers) => {
+    const className = showShape
+        ? styles.legendItemLineWithIcon
+        : styles.legendItemLine;
+    const svg = itemPath.append("svg").attr("class", className);
```
Instead of `.attr("class", className)` you can use `.classed`

```js
itemPath
    .append("svg")
    .classed(
        showShape ? styles.legendItemLineWithIcon : styles.legendItemLine,
        true
    );


```

abhijit945 commented on src/main/js/helpers/legend.js:
```diff
@@ -92,27 +92,85 @@ const loadLegendItem = (legendSVG, t, shownTargets, eventHandlers) => {
+        .attr("y1", 1.5)
+        .attr("y2", 1.5)
+        .attr(
+            "style",
```
I see we are creating 2 lines for sole purpose of getting a white stroke. Can we have the css classes on the lines instead of SVG? That way we can remove some of these inline styles.

abhijit945 commented on src/main/js/helpers/styles.js:
```diff
@@ -38,6 +38,10 @@ export default {
+    legendItemIconWithLine: "carbon-legend-item-icon-with-line",
+    legendItemLine: "carbon-legend-item-line",
+    legendItemLineWithIcon: "carbon-legend-item-line-with-icon",
+    legendItemSpan: "carbon-legend-item-span",
```
What's the difference between `legendItemIconWithLine` and `legendItemLineWithIcon`?

abhijit945 commented on src/main/less/core.less:
```diff
@@ -102,7 +102,9 @@ svg:not(:root) {
             align-content: center;
             margin-right: 0.5rem;
-            padding: 0.5rem;
+            padding: 0;
```
Are you sure you dont want any padding?

abhijit945 commented on src/main/less/core.less:
```diff
@@ -138,10 +140,39 @@ svg:not(:root) {
+
+        & .carbon-legend-item-line {
+            display: block;
+            stroke: @dark;
```
Isnt this black? Is this what we want?

abhijit945 commented on src/main/js/helpers/legend.js:
```diff
@@ -92,27 +92,85 @@ const loadLegendItem = (legendSVG, t, shownTargets, eventHandlers) => {
+    const svg = itemPath.append("svg").attr("class", className);
+    svg.append("line") // creating white line
+        .attr("x1", 0)
+        .attr("x2", 35)
```
Move magic number to `constants.js`.

abhijit945 commented on dev/app.js:
```diff
@@ -207,6 +208,11 @@ renderSiteApp(
+                {
+                    pathname: "/line/legend-options",
+                    content: renderLegendWithOptions,
+                    title: "Custom Legend Options"
```
`Line with legend options`

SharathAlur replied:
Updated

abhijit945 commented on dev/examples/controls/line.js:
```diff
@@ -523,3 +523,11 @@ export const renderDashedLine = (id) => {
     lineDefault.loadContent(Carbon.api.line(axisData.data[1]));
     return lineDefault;
 };
+export const renderLegendWithOptions = (id) => {
```
`renderLineWithLegendOptions` 
Lets make this consistent with others.

SharathAlur replied:
Done

abhijit945 commented on src/main/js/helpers/legend.js:
```diff
@@ -92,27 +94,100 @@ const loadLegendItem = (legendSVG, t, shownTargets, eventHandlers) => {
+ * @param {object} buttonPath - d3 svg object
+ * @param {object} t - input item object processed from the input JSON
+ */
+const processLegendOptions = (buttonPath, t) => {
```
Can we change the variable `t`? In my previous comment I think I named this something else.

SharathAlur replied:
Updated to input

abhijit945 commented on src/main/js/helpers/legend.js:
```diff
@@ -92,27 +94,100 @@ const loadLegendItem = (legendSVG, t, shownTargets, eventHandlers) => {
+ * @param {object} t - input item object processed from the input JSON
+ */
+const createLegendIcon = (buttonPath, t) => {
+    buttonPath.append(() =>
```
Remove the braces since you are saying that you are returning an object in JSDoc

SharathAlur replied:
Updated

abhijit945 commented on src/main/js/helpers/legend.js:
```diff
@@ -92,27 +94,100 @@ const loadLegendItem = (legendSVG, t, shownTargets, eventHandlers) => {
+                ? constants.DEFAULT_LEGEND_LINE_WIDTH_WITH_SYMBOL - 1
+                : constants.DEFAULT_LEGEND_LINE_WIDTH - 1
+        )
+        .attr("y1", 1.5)
```
Move this value to `constants.js`

SharathAlur replied:
Moved

abhijit945 commented on src/main/js/helpers/legend.js:
```diff
@@ -290,6 +365,26 @@ const loadPieLegendItem = (legendSVG, dataTarget, { hoverHandler }) => {
+ */
+const getDefaultLegendOptions = (graphConfig, dataTarget) => {
+    const legendOptions = getDefaultValue(dataTarget.legendOptions, {
+        showShape: getDefaultValue(graphConfig.showShapes, true),
```
So we are connecting the showShapes? Didnt we decide against this to not impact current consumers?

SharathAlur replied:
I have just set showShape to true to avoid the impact on the existing customers.

abhijit945 commented:
:tada: This PR is included in version 1.22.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/@cerner/carbon-graphs/v/1.22.0)
- [GitHub release](https://github.com/cerner/carbon-graphs/releases/tag/v1.22.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
